## Notebook is summary of the experiment results ##

Two way to view tensorboard:

1. To start a TensorBoard session from VSC:

    1. Select the Python Interpreter for the notebook. You can select python interpreter from Command Palette -> Python Select Interpreter
    2. Open the command palette (Ctrl/Cmd + Shift + P)
    3. Search for the command “Python: Launch TensorBoard” and press enter.
    4. You will be able to select the folder where your TensorBoard log files are located. By default, the current working directory will be used.

  VSCode will then open a new tab with TensorBoard and its lifecycle will be managed by VS Code as well.

2. From command prompt: 
    
    1. Open terminal

        > log_path = 'logs' #Based on the folder in the terminal this will change. I was always on parent dir so provide relative path</br>

        > tensorboard --logdir log_path --host localhost --port 8888
        
            OR
        
        > tensorboard --logdir 'logs' --host localhost --port 8888

    2. Open web browser and type in URL
        http//localhost:8888

Note:
1. Move up to directory 
     log_path = os.path.join(os.getcwd() ,"..\\logs")#Since in my case to reach log folder have to move 1 up to reach parent director. So used single '..\\' 
2. https://stackoverflow.com/questions/63938552/how-to-run-tensorboard-in-vscode


**List of selected hyperparam with gave decent F1-score result**

<table>
    <tr>
        <td>test name</td>
        <td>lr</td>
        <td>lstm unit</td>
        <td>dropout</td>
        <td>recurent dropout</td>
        <td>l2</td>
        <td>batch size</td>
        <td>epoch</td>
        <td>f1-score </td>
        <td></td>
        <td>remark </td>
    </tr>
    <tr>   
        <td> </td>
        <td> </td>
        <td> </td>
        <td> </td>
        <td> </td>
        <td> </td>
        <td> </td>
        <td> </td>
        <td>train</td>
        <td>valid</td>
        <td> </td>
    </tr>
    <tr>
        <td>Bidirection LSTM</td>
        <td>0.01</td>
        <td>256</td>
        <td>0.2</td>
        <td>0.2</td>
        <td>default</td>
        <td>2048</td>
        <td>150</td>
        <td>0.48</td>
        <td>0.42</td>
        <td>Overfitting. Best value train: 0.43, Valid:0.425, Epoch:85 </td>
    </tr>
    <tr>
        <td>Trainable Weighted bi-direction LSTM</td>
        <td>0.002</td>
        <td>256</td>
        <td>0.2</td>
        <td>0.2</td>
        <td>default</td>
        <td>512</td>
        <td>41</td>
        <td>0.61</td>
        <td>0.49</td>
        <td>Overfitting. Best value train: 0.4165, Valid:0.4001, Epoch:20 </td>
    </tr>
    <tr>
        <td>Trainable bi-direction LSTM</td>
        <td>0.002</td>
        <td>256</td>
        <td>0.2</td>
        <td>0.2</td>
        <td>default</td>
        <td>512</td>
        <td>25</td>
        <td>0.56</td>
        <td>0.48</td>
        <td>Overfitting. Best value train: 0.3467, Valid:0.3392, Epoch:16 </td>
    </tr>
    <tr>
        <td>Weighted LSTM</td>
        <td>0.002</td>
        <td>128</td>
        <td>0.2</td>
        <td>0.1</td>
        <td>0.0003</td>
        <td>256</td>
        <td>300</td>
        <td>0.5901</td>
        <td>0.4816</td>
        <td>Overfitting. Best value train: 0.4462, Valid:0.4319, Epoch:115 </td>
    </tr>
    <tr>
        <td>Bert</td>
        <td>1.00E-05</td>
        <td>256</td>
        <td>0.1</td>
        <td>-</td>
        <td>-</td>
        <td>19*8 (TPU)</td>
        <td>250</td>
        <td>0.5477</td>
        <td>0.5098</td>
        <td>Tensor size issue so kept the batch size smaller. Overfitting. Best value train: 0.5088, Valid:0.5022, Epoch:16 </td>
    </tr>
    <tr>
        <td>Bert</td>
        <td>1.00E-05</td>
        <td>256</td>
        <td>0.2</td>
        <td>-</td>
        <td>-</td>
        <td>64*8 (TPU)</td>
        <td>16</td>
        <td>0.2646</td>
        <td>0.3072</td>
        <td></td>
    </tr>
    <tr>
        <td>Bert</td>
        <td>1.00E-05</td>
        <td>256</td>
        <td>0.2</td>
        <td>-</td>
        <td>-</td>
        <td>64*8 (TPU)</td>
        <td>32</td>
        <td>0.4973</td>
        <td>0.4995</td>
        <td></td>
    </tr>
    <tr>
        <td>Bert</td>
        <td>1.00E-05</td>
        <td>256</td>
        <td>0.2</td>
        <td>-</td>
        <td>-</td>
        <td>64*8 (TPU)</td>
        <td>50</td>
        <td>0.6795</td>
        <td>0.508</td>
        <td>Overfitting on training</td>
    </tr>
    <tr>
        <td>Bert</td>
        <td>1.00E-05</td>
        <td>256</td>
        <td>0.2</td>
        <td>-</td>
        <td>-</td>
        <td>20*8 (TPU)</td>
        <td>16</td>
        <td>0.5261</td>
        <td>0.5078</td>
        <td></td>
    </tr>
    <tr>
        <td>Bert</td>
        <td>1.00E-05</td>
        <td>256</td>
        <td>0.2</td>
        <td>-</td>
        <td>-</td>
        <td>19*8 (TPU)</td>
        <td>15</td>        
        <td>0.5044</td>
        <td>0.5039</td>
        <td>Reasonable. Will use this model for Evaluating</td>
    </tr>
    <tr>
        <td>Bert</td>
        <td>4.00E-05</td>
        <td>210</td>
        <td>0.2</td>
        <td>-</td>
        <td>-</td>
        <td>64*8 (TPU)</td>
        <td>50</td>
        <td>0.7637</td>
        <td>0.5012</td>
        <td>Overfitting.  Best value train: 0.5234, Valid:0.5049, Epoch:11 </td>
    </tr>
</table>

## Eval the model on the test dataset ##

In [1]:
from src.utils.path_utils import get_project_root

#from pathlib import Path
import pandas as pd
import pickle
#import os

In [2]:
from src.config.configuration import ConfigurationManager
from src.entity.data_ingestion import DataIngestionConfig
from src.entity.data_cleaning import DataCleaningConfig

config = ConfigurationManager() #ConfigurationManager(notebook_artifacts = True)

model_train_artifact_config = config.get_model_artifacts_config()
data_ingestion_config = config.get_data_ingestion_config()
data_cleaning_config = config.get_data_cleaning_config()

sample_test_filename = 'SO_Tag_prediction_test.parquet'
sample_data_save_path = get_project_root().joinpath(data_ingestion_config.artifacts_dir).joinpath(data_ingestion_config.root_dir).joinpath(sample_test_filename)

#Sample from the test file
test_filename = get_project_root().joinpath(data_ingestion_config.artifacts_dir).joinpath(data_ingestion_config.root_dir).joinpath(data_ingestion_config.data_test_filename)

data_ingestion_config = DataIngestionConfig(artifacts_dir = data_ingestion_config.artifacts_dir,
                                            root_dir = data_ingestion_config.root_dir,  
                                            data_downlad_url = data_ingestion_config.data_downlad_url,
                                            zip_data = data_ingestion_config.zip_data,
                                            data_filename = sample_test_filename,
                                            data_test_filename = data_ingestion_config.data_test_filename
)

data_cleaning_config = DataCleaningConfig(artifacts_dir = data_cleaning_config.artifacts_dir,
                                          root_dir = data_cleaning_config.root_dir,
                                          selected_col = data_cleaning_config.selected_col,
                                          tag_col = data_cleaning_config.tag_col,
                                          cleaned_data_filename = 'test_tags_text_processed.parquet'
)

#del [data_cleaning_config, data_ingestion_config]

print()
print(data_ingestion_config)
(sample_data_save_path, test_filename)

[2024-01-06 16:24:48,173: INFO: common: yaml file: F:\github_workspace\stackoverflow_tag_question\config\config.yaml loaded successfully]
[2024-01-06 16:24:48,308: INFO: common: yaml file: F:\github_workspace\stackoverflow_tag_question\params.yaml loaded successfully]
[2024-01-06 16:24:48,341: INFO: common: yaml file: F:\github_workspace\stackoverflow_tag_question\schema.yaml loaded successfully]

DataIngestionConfig(artifacts_dir='artifacts', root_dir='data_ingestion', data_downlad_url='https://www.kaggle.com/competitions/stackoverflow-moderation/data', zip_data='stackoverflow-moderation.zip', data_filename='SO_Tag_prediction_test.parquet', data_test_filename='test.parquet')


(WindowsPath('F:/github_workspace/stackoverflow_tag_question/artifacts/data_ingestion/SO_Tag_prediction_test.parquet'),
 WindowsPath('F:/github_workspace/stackoverflow_tag_question/artifacts/data_ingestion/test.parquet'))

### Sample and Clean test dataset ###

In [ ]:
dbset = pd.read_parquet(test_filename)
dbset = dbset.sample(100, random_state = 42)
dbset.head()

In [ ]:
dbset['BodyMarkdown'] = ""
dbset[['Tag1', 'Tag2', 'Tag3', 'Tag4','Tag5']]  = dbset.tags.str.split(' ', expand = True) #.apply(lambda x: x.split()[:5])
dbset.fillna("", inplace = True)

dbset.rename(columns = {'question':'Title'}, inplace = True)
dbset.drop(columns = ['tags'], inplace = True)
dbset.head()

dbset.to_parquet(sample_data_save_path, engine = 'fastparquet')

In [ ]:
from src.entity.data_validation import DataValidation

data_validation_config = config.get_data_validation_config()
data_validation_obj = DataValidation(Config = data_validation_config, 
                                     Data_Config = data_ingestion_config,
                                    )

data_validation_obj.validate_all_columns()


In [ ]:
from src.entity.data_cleaning import DataCleaning

data_cleaning_obj = DataCleaning(Config = data_cleaning_config, 
                                 Data_Config = data_ingestion_config,
                                )

data_cleaning_obj.cleaning()

###  Validate on test data ###

In [3]:
test_filename = get_project_root().joinpath(data_cleaning_config.artifacts_dir).joinpath(data_cleaning_config.root_dir).joinpath(data_cleaning_config.cleaned_data_filename)

test_filename

WindowsPath('F:/github_workspace/stackoverflow_tag_question/artifacts/data_cleaning/test_tags_text_processed.parquet')

In [4]:
dbset = pd.read_parquet(test_filename)
dbset.head()

tags  \
0                         javascript,asp.net,textbox   
1             jquery,jquery-ui,slider,jquery-animate   
2                                       uiscrollview   
3  titanium,appcelerator,getjson,appcelerator-mob...   
4                                 javascript,raphael   

                                               title  
0  add color palett aspnet textbox add color pale...  
1  anim slider handl alter jqueri ui slider valu ...  
2  algorithm calcul uiscrollview landscap content...  
3  refresh dont append data refresh dont append d...  
4  attach bind text element raphael attach bind t...

#### Steps: Prepare data ####

1. Convert comma seperated tags into onehot encoded columns as Y labels
2. Take selected features name from Y labels
3. Text tokenizer

In [5]:
foldername = "train_bert"
serialize_objects_dir = get_project_root().joinpath(model_train_artifact_config.artifacts_dir).joinpath(model_train_artifact_config.serialize_objects_dir).joinpath(foldername)
serialize_multilabel_binarizer_obj_dir = serialize_objects_dir.joinpath("multilabel_binarizer_obj.pickle")
#serialize_multilabel_binarizer_obj_dir

#Step 1: Load object to convert the comma seperated tags into Y labels

#Check if the file not found throw exception
if serialize_multilabel_binarizer_obj_dir.exists():

       #Load multi label binarizer object
       with open(serialize_multilabel_binarizer_obj_dir,'rb') as file:
              multilabel_binarizer_obj = pickle.load(file)

f:\github_workspace\stackoverflow_tag_question\.env\lib\site-packages\sklearn\base.py:347: InconsistentVersionWarning: Trying to unpickle estimator MultiLabelBinarizer from version 1.2.2 when using version 1.3.0. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations
  warnings.warn(


In [6]:
#Convert the tags
Y = multilabel_binarizer_obj.transform(dbset.tags.apply(lambda x: { i for i in x.split(',')}))

print(f'Y label shape: {Y.shape}')
Y

Y label shape: (100, 14878)


f:\github_workspace\stackoverflow_tag_question\.env\lib\site-packages\sklearn\preprocessing\_label.py:900: UserWarning: unknown class(es) ['android-loadermanager', 'approximation-hardness', 'candidate', 'cc.complexity-theory', 'composite-component', 'dynamical-systems', 'fastreport', 'function-calls', 'google-api-python-client', 'hcard', 'icacls', 'jquery-animate', 'mpm-prefork', 'retina-display', 'sharepoint-enterprise', 'tortoise-svn', 'window-manager', 'windows-search', 'word-processing'] will be ignored
  warnings.warn(


array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [7]:
#check if all rows of column is empty/null/zero
arr = Y.sum(axis = 0)
non_zero_idx = arr.nonzero()
non_zero_idx

(array([   13,   163,   186,   193,   194,   223,   225,   246,   315,
          342,   374,   464,   483,   520,   598,   654,   681,   682,
          703,   727,   783,   797,   836,   846,   880,   964,   999,
         1047,  1095,  1107,  1171,  1211,  1259,  1260,  1373,  1374,
         1484,  1501,  1608,  1609,  1611,  1658,  1679,  1699,  1829,
         1929,  2035,  2091,  2253,  2268,  2645,  2745,  2758,  2917,
         2930,  2938,  2977,  3092,  3171,  3180,  3219,  3438,  3645,
         3724,  3800,  3966,  4060,  4130,  4134,  4143,  4148,  4149,
         4153,  4157,  4226,  4251,  4331,  4452,  4501,  4703,  4889,
         4898,  4969,  5018,  5066,  5167,  5213,  5223,  5236,  5413,
         5503,  5748,  5774,  5790,  5794,  5996,  6016,  6368,  6375,
         6399,  6548,  6599,  6605,  6624,  6782,  6820,  6833,  6862,
         6864,  6884,  6989,  7066,  7174,  7319,  7334,  7478,  7496,
         7521,  7828,  7925,  7947,  8081,  8386,  8392,  8412,  8700,
      

In [8]:
len(non_zero_idx[0])

209

**Observation:**

With this test dataset outof 14878 label only 209 label are associate to some records. Other label are all zeros i.e there are no records belong to those label.

In [9]:
#Check if each records have atleast 1 label assigned to it
Y.sum(axis = 1)

array([3, 3, 1, 5, 2, 2, 2, 2, 4, 3, 3, 5, 2, 5, 5, 4, 1, 2, 1, 3, 5, 1,
       4, 2, 4, 4, 3, 2, 3, 3, 5, 3, 2, 3, 3, 0, 3, 2, 1, 2, 4, 3, 2, 1,
       4, 4, 1, 2, 3, 3, 5, 0, 4, 5, 4, 5, 1, 2, 3, 3, 2, 2, 3, 3, 4, 5,
       2, 3, 4, 1, 3, 2, 4, 2, 3, 3, 3, 2, 3, 3, 4, 4, 3, 4, 1, 2, 1, 3,
       3, 3, 1, 1, 5, 4, 4, 3, 5, 4, 4, 1])

**Observartion:**

There are records where does not have any label. We shall drop those records

In [10]:
#Step 2: Load list of selected features name
serialize_selected_features_names_dir = serialize_objects_dir.joinpath("selected_features_names.pickle")

#Check if the file not found throw exception
if serialize_selected_features_names_dir.exists():

       #Open and read the file
       with open(serialize_selected_features_names_dir,'rb') as file:
              selected_features_name = pickle.load(file)

In [11]:
selected_features_name[0]

'php'

In [12]:
len(selected_features_name)

185

In [13]:
len(multilabel_binarizer_obj.classes_)

14878

In [14]:
%%time

#multilabel_binarizer_obj.classes_ is array object. 
def find_indices(list_to_check, item_to_find):
    return [idx for idx, value in enumerate(list_to_check) if value == item_to_find ] # Option 2. This take  4.5 ms
    #return np.where(multilabel_binarizer_obj.classes_ == item_to_find) # Option 1: This take 171ms

selected_tags_idx = [ find_indices(multilabel_binarizer_obj.classes_, item_to_find)[0] for item_to_find in selected_features_name ]

CPU times: total: 328 ms
Wall time: 411 ms


In [15]:
#Take the selected columns
Y_test = Y[:,selected_tags_idx]
Y_test.shape

(100, 185)

In [16]:
## Step 3: Removing the question with no tags selected

import numpy as np

row_idx = np.nonzero(Y_test.sum(axis = 1))[0] #Question with nonzero tags
Y_test = Y_test[row_idx,:]
X_test = dbset.iloc[row_idx,:]

X_test.shape

(76, 2)

In [17]:
import gc

del dbset
gc.collect()

0

In [18]:
from src.train_scripts.model_bert import model_hyperparam, model_tokenizer, model_import_tokenizer

## Step 4: Int tokenizer on text 

#Load the tokenizer
#MODEL_NAME = 'bert-large-uncased'
#hyper_param = {}
#hyper_param['model_name'] = MODEL_NAME
hyper_param = model_hyperparam()

tokenizer = model_tokenizer(hyper_param) 

#serialize_tokenizer_obj_dir = serialize_objects_dir.joinpath("tokenizer_obj.pickle")
#tokenizer = model_import_tokenizer(serialize_tokenizer_obj_dir)

f:\github_workspace\stackoverflow_tag_question\.env\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


[2024-01-06 16:28:40,361: INFO: common: yaml file: F:\github_workspace\stackoverflow_tag_question\config\config.yaml loaded successfully]
[2024-01-06 16:28:40,377: INFO: common: yaml file: F:\github_workspace\stackoverflow_tag_question\params.yaml loaded successfully]
[2024-01-06 16:28:40,384: INFO: common: yaml file: F:\github_workspace\stackoverflow_tag_question\schema.yaml loaded successfully]


In [ ]:
from src.train_scripts.distributed_train import define_distributed_strategy

strategy, num_replicas_in_sync = define_distributed_strategy()

In [19]:
#Step 7: Create the dataset
from src.train_scripts.model_bert import model_encode
import tensorflow as tf

test_batch = tf.data.Dataset.from_tensor_slices(X_test).map(lambda x: model_encode(x, tokenizer, hyper_param['max_title_len']), num_parallel_calls = tf.data.experimental.AUTOTUNE).prefetch(tf.data.AUTOTUNE) 

test_batch

<PrefetchDataset element_spec={'input_ids': TensorSpec(shape=(None, 80), dtype=tf.int64, name=None), 'attention_mask': TensorSpec(shape=(None, 80), dtype=tf.int64, name=None)}>

#### Steps: Load model ####

In [ ]:
from src.train_scripts.model_bert import  define_model

with strategy.scope(): 

    model = define_model(hyper_param)
    model.compile()
    model.summary()

In [ ]:
#Step 8: Load the model

model_artifacts = config.get_model_artifacts_config()
checkpoint = "logs_hparam_lr1e-05_units256_dropout02_batch19_epoch15" 
checkpoint_dir = get_project_root().joinpath(model_artifacts.artifacts_dir).joinpath(model_artifacts.checkpoint_dir).joinpath(checkpoint)
checkpoint_file = checkpoint_dir.joinpath('best_model.weights.h5')

with strategy.scope(): 
    model.load_weights(checkpoint_file)

#### Steps: Predict ####

In [22]:
import math

threshold = 0.60 

def sigmoid(x):

    sig =  1 / (1 + math.exp(-x))    
    #return 1 if sig > threshold else 0
    return 1 if x > threshold else 0

In [23]:
#Step 9: Predict the model

y_hat = model.predict(test_batch)

76/76 [==============================] - 213s 2s/step


In [25]:
index = 5

print(f'Y_label index: {np.nonzero(Y_test[index])}, label: {", ".join(selected_features_name[np.nonzero(Y_test[index])])}')
print(f'Y_predict index : {np.nonzero(list(map(sigmoid, y_hat[index])))}, label: {", ".join(selected_features_name[np.nonzero(list(map(sigmoid, y_hat[index])))])}')

Y_label index: (array([8], dtype=int64),), label: python
Y_predict index : (array([  0,   2,   3,   4,   5,   6,   7,  10,  12,  13,  14,  15,  19,
        20,  23,  24,  26,  30,  31,  32,  34,  38,  40,  41,  44,  46,
        48,  50,  53,  55,  56,  57,  58,  59,  60,  62,  63,  64,  66,
        67,  70,  71,  72,  73,  74,  75,  79,  80,  81,  82,  83,  84,
        85,  86,  90,  91,  92,  93,  94,  98,  99, 100, 101, 103, 105,
       107, 108, 109, 110, 111, 113, 114, 115, 116, 119, 120, 123, 125,
       129, 132, 133, 134, 135, 136, 137, 138, 139, 140, 142, 144, 145,
       146, 147, 148, 150, 151, 153, 154, 156, 157, 159, 162, 163, 164,
       165, 167, 168, 169, 174, 175, 176, 178, 179, 181, 183], dtype=int64),), label: php, c#, javascript, android, c++, jquery, iphone, mysql, .net, c, css, sql, ruby-on-rails, database, ruby, algorithm, sql-server, books, regex, html5, wpf, performance, apache, eclipse, image, web-services, php5, oop, visual-studio-2010, programming-languages, 